keps/sig-storage/20190408-volume-scheduling-limits.md

```diff
@@ -101,26 +108,25 @@ It is problematic for CSI (`MaxCSIVolumeCountPred`) outlined in [#730](https://g
     * If a CSI driver is registered for an in-tree volume plugin and it reports a different volume limit than in-tree volume plugin, the limit reported by CSI driver is used and kubelet logs a warning.
   * User may NOT change `CSINode.status.allocatable` to override volume plugin / CSI driver values, e.g. to "reserve" some attachment to the operating system. Kubelet will periodically reconcile `CSINode` and overwrite the value.
-    * Especially, `kubelet --kube-reserved` or `--system-reserved` cannot be used to "reserve" volumes for kubelet or the OS. It is not possible with current kubelet and this KEP does not change it.
-      * We expect that CSI drivers will have configuration options / cmdline arguments to reserve some volumes and they will report their limit already reduced by that reserved amount. 
+    * Especially, `kubelet --kube-reserved` or `--system-reserved` cannot be used to "reserve" volumes for kubelet or the OS. It is not possible with current kubelet and this KEP does not change it. We expect that CSI drivers will have configuration options / cmdline arguments to reserve some volumes and they will report their limit already reduced by that reserved amount.
 
   * Kubelet will continue filling `Node.status.allocatable` and `Node.status.capacity` for both in-tree and CSI volumes during deprecation period. After the deprecation period, it will stop using them completely.
     * Scheduler (all its storage predicates) will ignore `Node.status.allocatable` and `Node.status.capacity` if `CSINode.status.allocatable` is present.
     * If `CSINode.status.allocatable` (or whole `CSINode`) is missing, scheduler falls back to `Node.status.allocatable`. This solves version skew between old kubelet (using `Node.status`) and new scheduler.
     * After deprecation period, scheduler won't schedule any pods that use volumes to a node with missing `CSINode` instance. It is expected that it happens only during node registration when `Node` exists and `CSINode` doesn't and it self-heals quickly.
 
-* `CSINode.status.allocatable` is a map CSI Driver name -> int64. Following combinations are possible:
+* `CSINode.status.allocatable` is array of limits. Following combinations are possible:
 
-| Driver name       | value | Description  |
-| ----------------- | ----- | ------------ |
-| `ebs.csi.aws.com` | 0     | plugin / CSI driver exists and has zero limit, i.e. can attach no volumes |
-| `ebs.csi.aws.com` | X>0   | plugin / CSI driver exists and can attach X volumes (where X > 0) |
-| `ebs.csi.aws.com` | X<0   | negative values are blocked by validation
-| key is missing in `CSINode.status.allocatable`   | -     |  there is no limit of volumes on the node* |
+|  `Volumes` | Description  |
+|  --------- | ------------ |
+|  0         | plugin / CSI driver exists and has zero limit, i.e. can attach no volumes |
```

Comment:
Similar question comes up if in the future we add capacity limits. How to tell the difference between 0 and undefined?  Do we need the fields to be pointers?

Reply:
While discussing the issue of stopping pods from getting scheduled on a node if driver is not installed on it - I thought it might be best to handle this as a separate issue. We can indeed set value to `0` when driver is not installed but this applies only to drivers that support migration. 

Since we won't be creating `CSINode` object for in-tree drivers which don't support migration the solution of setting the volume limit to `0` might not be good enough. We may have to go one step further and say prevent scheduling of volumes if `CSINode` object for a driver does not exist on a node? If yes then - that sounds like something that is kinda out of scope for volume limit work.

Reply:
I'm only concerned about drivers that are being migrated.  If we set the limit to "unlimited", then during this (hopefully) brief moment, Pods that are using in-tree drivers can fail attaching, and require manual intervention to fix.  One of the goals of migration is that users shouldn't notice that we switched.

Reply:
> Similar question comes up if in the future we add capacity limits. How to tell the difference between 0 and undefined?

If a driver is installed, it is in `CSINode.Spec.Drivers`. Then the driver either has its limit in `CSINode.Status` or it's missing there, and that means there is no limit.

IMO, scheduling based on on availability of a driver on a node is a separate issue and separate predicate. IMO it should be fairly trivial to implement.

Reply:
It could be a separate issue but we also need to decide if that should block migration ga or not. @davidz627

Reply:
Scheduling based on availability of a driver on a node is a blocker for migration GA since not having that would be a regression in volume behavior.

Reply:
This was addressed by @jsafrane in recent commit. Scheduler will not schedule pods to a node if it does not have driver entry in `CSINode` object.